### PR TITLE
Refactor GatewayClient retry mechanism

### PR DESCRIPTION
**Pull request by omerfirmak**
Old implementation was not quite doing what it was supposed to

### CODE REVIEW

File: clients/gateway_test.go [hunk 1]
```diff
@@ -517,11 +517,15 @@ func TestGetClassDefinition(t *testing.T) {
 }
 
 func TestHttpError(t *testing.T) {
+	maxRetries := 2
+	callCount := make(map[string]int)
 	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		soFar := callCount[r.URL.String()]
+		callCount[r.URL.String()] = soFar + 1
```
Comment (IronGauntlets) [inline clients/gateway_test.go:524]:
```suggestion
		callCount[r.URL.String()]++
```

Not entirely sure the above will work but worth a try.

**Review by IronGauntlets — APPROVED** [review]

File: clients/gateway.go [hunk 2]
```diff
@@ -87,32 +86,35 @@ func (c *GatewayClient) buildQueryString(endpoint string, args map[string]string
...
 		}
 	}
-	if err != nil {
-		return nil, err
-	} else if res.StatusCode != http.StatusOK {
-		return nil, errors.New(res.Status)
-	}
-
-	return io.ReadAll(res.Body)
+	return nil, err
```
Comment (joshklop) [inline clients/gateway.go:117]: Will this return a nil error after `maxRetries`? If so, maybe we can create a new error?

File: clients/gateway_test.go (hunk 1 above)
Comment (omerfirmak) [reply clients/gateway_test.go]: works:) thanks!

**Review by joshklop — APPROVED** [review]

File: clients/gateway.go (hunk 2 above)
Comment (omerfirmak) [reply clients/gateway.go:117]: It just returns the last error `Do` returned. am I missing something?

Comment (IronGauntlets) [reply clients/gateway.go:117]: I believe it will the error on line 105

Comment (omerfirmak) [reply clients/gateway.go:117]: or what L105 sets it to

**Pull request merged**